### PR TITLE
added integration tests

### LINKED ISSUE #16: [Core]: Add integration tests
Add integration tests for controllers in the [core](https://github.com/5GSEC/nimbus/tree/main/internal/controller) to verify the NimbusPolicy and ClusterNimbusPolicy generation.

#### Create and Read

- **Scenario**: Create a SecurityIntent and a SecurityIntentBinding.

- **Expected outcomes:**
  - [x] An intermediary NimbusPolicy is generated with the expected values populated correctly.

  - [x] The created SecurityIntent status subresource includes the ID and action fields with the intent values.
The created SecurityIntentBinding status subresource reflects the number and names of bound intents, along with the generated NimbusPolicy name.

- [ ] Similar test cases for a SecurityIntent with a ClusterSecurityIntentBinding, expecting a ClusterNimbusPolicy generation.

#### Update
 - **Scenario**: Modify a SecurityIntentBinding

 - **Expected Outcome**:
    - [x] The corresponding NimbusPolicy is updated.

- **Scenario**: Modify a NimbusPolicy

- **Expected Outcome**:
   - [x] The NimbusPolicy reverts its changes.

- [ ] Similar test cases for a ClusterSecurityIntentBinding with corresponding ClusterNimbusPolicy update.

#### Delete

- **Scenario**: Delete a SecurityIntentBinding.

- **Expected Outcome**:
    - [x] The corresponding NimbusPolicy is deleted.

- [ ] Similar test case for a ClusterSecurityIntentBinding with corresponding ClusterNimbusPolicy deletion.

### PR DESCRIPTION
**Pull request by VedRatan**
## Description

 This PR added the required integration tests for NImbus to make code complaint with the tests,

Fixes #16 

**Does this PR introduce a breaking change?**

## Checklist

- [ ] PR title follows the `<type>: <description>` convention
- [ ] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
- [ ] I have updated the [documentation](../docs) accordingly
- [ ] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
- [ ] I have performed a self-review of my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/controllers/sib-deletion/chainsaw-test.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+            namespace: $(namespace)
+            name: dns-manipulation-binding
+          expect:
+          - match:
+              apiVersion: intent.security.nimbus.com/v1
+              kind: SecurityIntentBinding
+              namespace: $(namespace)
+              name: dns-manipulation-binding
+            check:
+              ($error != null): true
```
Comment (anurag-rajawat) [inline tests/controllers/sib-deletion/chainsaw-test.yaml:32]: Here you're only checking SIB deletion, we also need to check the deletion for corresponding NP.

File: tests/controllers/sib-deletion/sib.yaml
Comment (anurag-rajawat) [inline tests/controllers/sib-deletion/sib.yaml:1]: Why did you add this file if you're not using it?

File: tests/controllers/np-creation/chainsaw-test.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+apiVersion: chainsaw.kyverno.io/v1alpha1
+kind: Test
+metadata:
+  name: np-creation
+spec:
+  timeouts:
+    assert: 10s
+  steps:
+  - try:
+    # first operation: create the config map
```
Comment (anurag-rajawat) [inline tests/controllers/np-creation/chainsaw-test.yaml:10]: Looks like you forgot to update or remove the comment.

File: tests/controllers/np-creation/chainsaw-test.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    - apply:
+        # file is relative to the test folder
+        file: ../../../examples/namespaced/dns-manipulation-si-sib.yaml
+    # second operation: verify the config map exists and contains the expected data
```
Comment (anurag-rajawat) [inline tests/controllers/np-creation/chainsaw-test.yaml:14]: Looks like you forgot to update or remove the comment.

File: .github/workflows/pr-checks.yaml [hunk 4]
```diff
@@ -97,3 +97,52 @@ jobs:
...
         working-directory: ./pkg/adapter/${{ matrix.adapters }}
         run: make docker-build
+
+  chainsaw-tests:
+    runs-on: ubuntu-latest
+    steps:
+        - name: Checkout
+          uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+
+          # install tools
```
Comment (anurag-rajawat) [inline .github/workflows/pr-checks.yaml:107]: This comment is redundant.

File: .github/workflows/pr-checks.yaml [hunk 5]
```diff
@@ -97,3 +97,52 @@ jobs:
...
+        - name: Install helm
+          id: helm
+          uses: azure/setup-helm@5119fcb9089d432beecbf79bb2c7915207344b78 # v3.5
+          with:
+            token: ${{ secrets.GITHUB_TOKEN }}
+          
+          # install go
```
Comment (anurag-rajawat) [inline .github/workflows/pr-checks.yaml:114]: This comment is redundant.

File: .github/workflows/pr-checks.yaml [hunk 6]
```diff
@@ -97,3 +97,52 @@ jobs:
...
+
+        - name: Install go
+          uses: actions/setup-go@v5
+          with:
+            go-version: '1.21.7' # The Go version to download (if necessary) and use.
```
Comment (anurag-rajawat) [inline .github/workflows/pr-checks.yaml:119]: Install `1.21` go version, omit the patch version.

File: .github/workflows/pr-checks.yaml [hunk 7]
```diff
@@ -97,3 +97,52 @@ jobs:
...
+          
+
+        # install chainsaw
```
Comment (anurag-rajawat) [inline .github/workflows/pr-checks.yaml:122]: This comment is redundant.

File: .github/workflows/pr-checks.yaml [hunk 8]
```diff
@@ -97,3 +97,52 @@ jobs:
...
+
+        - name: Install chainsaw
+          uses: kyverno/action-install-chainsaw@8307cf9038333ce168dd4339ae24f9ab16ab7a00 # v0.1.4
+
+        # create kind cluster
```
Comment (anurag-rajawat) [inline .github/workflows/pr-checks.yaml:127]: This comment is redundant.

Comment (anurag-rajawat) [inline .github/workflows/pr-checks.yaml:127]: This comment is redundant.

File: .github/workflows/pr-checks.yaml [hunk 9]
```diff
@@ -97,3 +97,52 @@ jobs:
...
+
+        - name: Create k8s Kind Cluster
+          uses: helm/kind-action@v1
+
+
+        - name: start server
+          run: |
+            make install
+            make run &
```
Comment (anurag-rajawat) [inline .github/workflows/pr-checks.yaml:136]: Why do we need this?
Can't we use the following:
```shell
helm upgrade --install nimbus-operator . -n nimbus --create-namespace --set image.pullPolicy=Never
```
?

**Review by anurag-rajawat — CHANGES_REQUESTED** [review]

**Comment by anurag-rajawat** [comment]
We also need to add test cases to check the status of CRs.

File: .github/workflows/pr-checks.yaml (hunk 9 above)
Comment (VedRatan) [reply .github/workflows/pr-checks.yaml]: It's showing the  error while deploying nimbus via locally built image.

Comment (VedRatan) [reply .github/workflows/pr-checks.yaml]: image

**Review by shivaccuknox — APPROVED** [review]

**Review by shivaccuknox — APPROVED** [review]
